### DISCUSSION TITLE
SyncLogger in Crazyflie Python library returning identical states from different drones

### DISCUSSION THREAD

**Discussion by matteoviel18** [bitcraze/discussions#636] [General]
Hi everyone,

I'm working on a project where I'm using two Crazyflie drones and trying to log their states using the SyncLogger class provided by the cflib library. I've set up a dictionary state_loggers to store the loggers for each drone, and I'm using the state_logger_drone function to initialize and connect the loggers to each drone. Here's the relevant Python code (relevant parts only -  link to the full code below):

``` python
#State logger configuration
state_logger_config = LogConfig(name='state_conf', period_in_ms=200)
state_logger_config.add_variable('stateEstimate.x', 'float')
state_logger_config.add_variable('stateEstimate.y', 'float')
state_logger_config.add_variable('stateEstimate.z', 'float')
state_logger_config.add_variable('stateEstimate.vx', 'float')
state_logger_config.add_variable('stateEstimate.vy', 'float')
state_logger_config.add_variable('stateEstimate.vz', 'float')

#Instantiation of swarm:
swarm = Swarm(uris, factory=_Factory())

#Configuring the loggers
state_loggers_swarm()

def state_logger_drone(scf):
    state_logger = SyncLogger(scf, state_logger_config)
    state_logger.connect()
    state_loggers[scf._link_uri] = state_logger

def state_loggers_swarm():
    swarm.parallel_safe(state_logger_drone)
```

I then use the follow code to get the states:

``` python
for uri in uris:
     #Extracting information from the loggers
     state_data_dict[uri] = state_loggers[uri].next()
     #debug 
     print(uri, state_data_dict[uri]) 
     #ERROR: different drones have the same states
```

The problem I'm having is that the logged states for both drones are identical, even though they should be different, being two different drones. I'm not sure what I'm doing wrong.

If it would be helpful, you can find the full code available on GitHub:

https://github.com/matteoviel18/SyncLogger-for-swarm.git

where, using ROS, pace_state.py is used to trigger the state publication of my swarm at a certain pace, allowing to enter in the callback pace_sub_callback in my_swarm.py.

Can anyone help me figure out what's going on here? Any advice or suggestions would be greatly appreciated.

Thank you!
Matteo :)

**Comment by krichardsson** [comment]
The problem is that you reuse the `state_logger_config` object. It is not obvious, but the [`LogConfig`](https://github.com/bitcraze/crazyflie-lib-python/blob/9789d1a1cc269b22ca2e35bb466a09c33be8f3c5/cflib/crazyflie/log.py#L131) class keeps references to the Crazyflie object, callbacks and other stuff related to the logging. As you are using the same `state_logger_config` for both your Crazyflies, weird stuff will happen. It should work if you create a new instance of the config for each Crazyflie.

The design is far from good, but it goes back a long time and I blame the history :-) It should be re-written at some point.

**Reply by matteoviel18** [reply]
Dear Kristoffer, you are right: reusing state_logger_config gave problems.
Thank you for your help and prompt response: you solved my problem, amazing!
